lightning/src/routing/router.rs

```diff
@@ -1537,12 +1548,87 @@ where L::Target: Logger {
 	Ok(route)
 }
 
+// When an adversarial intermediary node observes a payment, it may be able to infer its
+// destination, if the remaining CLTV expiry delta exactly matches a feasible path in the network
+// graph. In order to improve privacy, this method obfuscates the CLTV expiry deltas along the
+// payment path by adding a randomized 'shadow route' offset to the final hop.
+fn add_random_cltv_offset(route: &mut Route, payment_params: &PaymentParameters, network: &NetworkGraph, random_seed_bytes: &[u8; 32]) {
+	let network_graph = network.read_only();
+	let network_channels = network_graph.channels();
+	let network_nodes = network_graph.nodes();
```

Comment:
I'm a little concerned that we lock these twice. Once here and once prior to this in `get_route`. It's possible the `NetworkGraph` could change in the interim. I think we want to call `add_random_cltv_offset` from within `get_route` instead such that a single read-only lock is used.

Reply:
I understand that locking twice is generally suboptimal. However, calling `add_random_cltv_offset` from `get_route` is probably not what we want, since we split them off to avoid having to make all dependent tests account for some random variety in the routes returned. This would not only entail a lot of changes, but also would require the testcases to be less strict. Maybe a middleground would be to lock the graph once in `find_route` and let `get_route` and `add_random_cltv_offset` take the returned `ReadOnlyNetworkGraph`?

Reply:
> Maybe a middleground would be to lock the graph once in `find_route` and let `get_route` and `add_random_cltv_offset` take the returned `ReadOnlyNetworkGraph`?

Yeah, I'm good with that. At one point `get_route` was the only public interface but we refactored it behind `find_route`.

Reply:
Great! Went this route with https://github.com/lightningdevkit/rust-lightning/pull/1286/commits/521ae09a03718c17cdafe52644213ce5c8fac134.